interfaces/builtin/fpga.go

```diff
@@ -0,0 +1,83 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const fpgaSummary = `allows access to the FPGA subsystem`
+
+const fpgaBaseDeclarationPlugs = `
+  fpga:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const fpgaBaseDeclarationSlots = `
+  fpga:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const fpgaConnectedPlugAppArmor = `
+# Description: Can access fpga subsystem.
+
+# Devices
+/dev/fpga[0-9]* rw,
+
+# /sys/class/fpga_* specified by:
+# https://github.com/torvalds/linux/blob/master/Documentation/ABI/testing/sysfs-class-fpga-manager
+# https://github.com/torvalds/linux/blob/master/Documentation/ABI/testing/sysfs-class-fpga-region
+# https://github.com/torvalds/linux/blob/master/Documentation/ABI/testing/sysfs-class-fpga-bridge
+/sys/class/fpga_manager/fpga[0-9]*/{name,state,status} r,
+/sys/class/fpga_region/region[0-9]*/compat_id r,
+/sys/class/fpga_bridge/bridge[0-9]*/{name,state} r,
+
+# Xilinx zynqmp FPGA, created by zynqmp_fpga_manager driver
+# https://github.com/torvalds/linux/blob/master/drivers/fpga/zynqmp-fpga.c
+/sys/devices/platform/firmware:zynqmp-firmware/firmware:zynqmp-firmware:pcap/fpga_manager/fpga[0-9]*/{name,state,status} r,
+/sys/devices/platform/firmware:zynqmp-firmware/firmware:zynqmp-firmware:pcap/fpga_manager/fpga[0-9]*/firmware w,
+/sys/devices/platform/firmware:zynqmp-firmware/firmware:zynqmp-firmware:pcap/fpga_manager/fpga[0-9]*/{flags,key} rw,
+/sys/devices/platform/fpga-full/fpga_region/region[0-9]*/compat_id r,
+
+# Xilinx zynqmp module parameters (not upstreamed yet)
+# https://github.com/Xilinx/linux-xlnx/blob/master/drivers/fpga/zynqmp-fpga.c#L36
+/sys/module/zynqmp_fpga/parameters/readback_type rw,
+
+# configfs interface
+/sys/kernel/config/device-tree/overlays/full{,1}/ rw,
+/sys/kernel/config/device-tree/overlays/full{,1}/path rw,
```

Comment:
I don't care for the device-tree access in this PR, especially after reading @alexmurray's assessment. There were arguments for adding something like this to the TPM interface and also, IIRC, for creating a generic device-tree interface and we said 'no'. While putting these accesses in a superprivileged interface (like the base decl is now) does protect against widespread abuse, looking at the accesses, there is nothing in there that says 'fpga' - they are just mostly generic accesses that allow changing anything so IMO, they don't belong in this interface.

I suggest we remove the device-tree overlays accesses and then make this interface a normal manually-connected interface instead of superprivileged. For the device-tree bits, I suggest one of:

1. leaving the device-tree stuff out of the interfaces and let device makers use gadget.yaml to program the fpga for the device and/or use a management snap (eg, in their brand store) that plugs system-files to program the device
2. we create a 'device-tree' superprivileged interface. I worry about this since AIUI changes to the device-tree could impact the operational status of the device in a way that cannot be undone by Ubuntu Core
3. we come up with some 'snap set system' options for fpga that gadget's and management snaps with snapd-control can use to have snapd make changes on the snap's behalf

'1' is my preferred short term solution. '2' I don't care for, but is possible. '3' is interesting on a theoretical level since allows a command to validate what will be overlayed, though reading a bit about FPGA, this seems impractical.

We need @pedronis to weigh in on the approach. IME, going with '1' is not controversial and would at least unblock device makers immediately.

Reply:
@jdstrand @alfonsosanchezbeato  1. seems fine to me if that's the security team recommendation. We should explore 3. at some point, it has appeared, been mentioned already more than a couple of times in different context, but I don't think any of the relevant teams have time ATM for it.

Reply:
@pedronis @jdstrand thanks for the review, I have removed now the permissions to load device tree overlays and converted this to a manually connected interface. I completely agree on that this generic access to overlays belongs elsewhere. Option 1 seems feasible to me.